File: frontend/src/types/vehicle.ts
/**
 * types/vehicle.ts — Interfaces véhicule (miroir du schema Pydantic backend).
 */

export type FuelType =
  | 'essence' | 'diesel' | 'hybride' | 'hybride_rechargeable'
  | 'electrique' | 'gpl' | 'hydrogene';

export type BodyType =
  | 'citadine' | 'berline' | 'suv' | 'break' | 'coupe'
  | 'cabriolet' | 'monospace' | 'utilitaire' | 'pick_up';

export type TransmissionType = 'manuelle' | 'automatique' | 'semi_auto';

export interface Vehicle {
  id: string;
  seller_id: string;
  brand: string;
  model: string;
  version?: string;
  year: number;
  mileage?: number;  // PIVOT: optional, always 0 for new vehicles
  fuel_type: FuelType;
  body_type: BodyType;
  transmission: TransmissionType;
  engine_power_hp?: number;
  color?: string;
  doors: number;
  seats: number;
  city: string;
  postal_code?: string;
  price: number;
  description?: string;
  status?: string;

  // Champs IA (nullable — remplis par les modules ML)
  predicted_price?: number;
  price_confidence?: number;
  condition_score?: number;   // ← PIVOT: dormant (vision module disabled for new cars)
  popularity_score?: number;  // ← Neo4j PageRank
  images?: Array<{ file_path: string }>;

  // Champs Catalogue Officiel
  trunk_volume_l?: number;
  ncap_rating?: string;
  fuel_consumption?: number;
  co2_emissions?: number;
  length_cm?: number;
  width_cm?: number;
  height_cm?: number;
  official_consumption?: number;
  real_consumption?: number;
  electric_range_km?: number;
  is_4x4?: boolean;
  engine_type?: string;
  condition?: string;
  source?: string;

  created_at: string;
  updated_at: string;

  // Relations imbriquées (selon l'endpoint)
  seller?: import('./user').UserBrief;
  source_url?: string;
}

export interface VehicleListResponse {
  items: Vehicle[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

export interface VehicleFilters {
  brand?: string;
  model?: string;
  city?: string;
  fuel_type?: FuelType | '';
  body_type?: BodyType | '';
  transmission?: TransmissionType | '';
  price_min?: number;
  price_max?: number;
  year_min?: number;
  year_max?: number;
  mileage_max?: number;
  doors?: number;
  seats?: number;
  color?: string;
  min_engine_power?: number;
  is_4x4?: boolean;
  condition?: 'neuf' | 'occasion' | string;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  group_by_model?: boolean;
  page?: number;
  page_size?: number;
}

/** Labels affichables pour les types de carburant */
export const FUEL_LABELS: Record<FuelType, string> = {
  essence: 'Essence',
  diesel: 'Diesel',
  hybride: 'Hybride',
  hybride_rechargeable: 'PHEV',
  electrique: '⚡ Électrique',
  gpl: 'GPL',
  hydrogene: 'Hydrogène',
};

/** Labels affichables pour les types de carrosserie */
export const BODY_LABELS: Record<BodyType, string> = {
  citadine: 'Citadine',
  berline: 'Berline',
  suv: 'SUV',
  break: 'Break',
  coupe: 'Coupé',
  cabriolet: 'Cabriolet',
  monospace: 'Monospace',
  utilitaire: 'Utilitaire',
  pick_up: 'Pick-up',
};

export const TRANSMISSION_LABELS: Record<TransmissionType, string> = {
  manuelle: 'Manuelle',
  automatique: 'Automatique',
  semi_auto: 'Semi-auto',
};

File: backend/app/models/vehicle.py
"""
models/vehicle.py — Modèle ORM Vehicle.
Correspond à la migration 002_create_vehicles.sql.
Champs IA (condition_score, predicted_price, popularity_score) nullable
pour l'instant — seront remplis par les modules vision, pricing, graphe.
"""

import uuid
from datetime import datetime, timezone

from sqlalchemy import (
    Boolean, DateTime, Enum, Integer, Numeric, SmallInteger, String, Text,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy import ForeignKey

from app.core.database import Base


class Vehicle(Base):
    __tablename__ = "vehicles"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    seller_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False
    )

    # ─── Identité ──────────────────────────────────────────────
    brand: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    model: Mapped[str] = mapped_column(String(100), nullable=False)
    version: Mapped[str | None] = mapped_column(String(200), nullable=True)
    year: Mapped[int] = mapped_column(Integer, nullable=False)
    mileage: Mapped[int] = mapped_column(Integer, nullable=False)
    fuel_type: Mapped[str] = mapped_column(
        Enum(
            "essence", "diesel", "hybride", "hybride_rechargeable",
            "electrique", "gpl", "hydrogene",
            name="fuel_type", create_type=False,
        ),
        nullable=False,
    )
    body_type: Mapped[str] = mapped_column(
        Enum(
            "citadine", "berline", "suv", "break", "coupe",
            "cabriolet", "monospace", "utilitaire", "pick_up",
            name="body_type", create_type=False,
        ),
        nullable=False,
    )
    transmission: Mapped[str] = mapped_column(
        Enum("manuelle", "automatique", "semi_auto", name="transmission_type", create_type=False),
        nullable=False,
        default="manuelle",
    )
    engine_power_hp: Mapped[int | None] = mapped_column(Integer, nullable=True)
    color: Mapped[str | None] = mapped_column(String(50), nullable=True)
    doors: Mapped[int] = mapped_column(SmallInteger, nullable=False, default=5)
    seats: Mapped[int] = mapped_column(SmallInteger, nullable=False, default=5)

    # ─── Caractéristiques Techniques Catalogue & Neuf ──────────
    trunk_volume_l: Mapped[int | None] = mapped_column(Integer, nullable=True)
    ncap_rating: Mapped[str | None] = mapped_column(String(50), nullable=True)
    fuel_consumption: Mapped[float | None] = mapped_column(Numeric(5, 2), nullable=True)
    co2_emissions: Mapped[float | None] = mapped_column(Numeric(6, 2), nullable=True)
    length_cm: Mapped[int | None] = mapped_column(Integer, nullable=True)
    is_4x4: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    engine_type: Mapped[str | None] = mapped_column(String(100), nullable=True)
    condition: Mapped[str] = mapped_column(String(50), nullable=False, default="new")
    source: Mapped[str] = mapped_column(String(100), nullable=False, default="wakala_catalogue")

    # ─── Statut (Soft Delete) ──────────────────────────────────
    status: Mapped[str] = mapped_column(
        Enum("available", "sold", "deleted", name="vehicle_status", create_type=False),
        nullable=False,
        default="available",
    )
    source_url: Mapped[str | None] = mapped_column(String(500), nullable=True)

    # ─── Localisation ──────────────────────────────────────────
    city: Mapped[str] = mapped_column(String(150), nullable=False, index=True)
    postal_code: Mapped[str | None] = mapped_column(String(10), nullable=True)
    latitude: Mapped[float | None] = mapped_column(Numeric(10, 7), nullable=True)
    longitude: Mapped[float | None] = mapped_column(Numeric(10, 7), nullable=True)

    # ─── Prix ──────────────────────────────────────────────────
    price: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    predicted_price: Mapped[float | None] = mapped_column(Numeric(12, 2), nullable=True)
    price_confidence: Mapped[float | None] = mapped_column(Numeric(5, 4), nullable=True)

    # ─── Scores IA (nullable — remplis par les modules ML) ────
    condition_score: Mapped[float | None] = mapped_column(Numeric(5, 2), nullable=True)
    popularity_score: Mapped[float | None] = mapped_column(Numeric(8, 4), nullable=True)

    # ─── Description ───────────────────────────────────────────
    description: Mapped[str | None] = mapped_column(Text, nullable=True)

    # ─── Timestamps ────────────────────────────────────────────
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=lambda: datetime.now(timezone.utc)
    )

    # ─── Relations ─────────────────────────────────────────────
    seller: Mapped["User"] = relationship("User", back_populates="vehicles", lazy="selectin")  # noqa: F821
    listings: Mapped[list["Listing"]] = relationship(  # noqa: F821
        "Listing", back_populates="vehicle", lazy="selectin"
    )

    saved_by_users: Mapped[list["User"]] = relationship(  # noqa: F821
        "User", secondary="saved_vehicles", back_populates="saved_vehicles", lazy="selectin"
    )
    reviews: Mapped[list["Review"]] = relationship(  # noqa: F821
        "Review", back_populates="vehicle", lazy="selectin"
    )
    services: Mapped[list["VehicleService"]] = relationship( # noqa: F821
        "VehicleService", back_populates="vehicle", lazy="selectin"
    )
    reminders: Mapped[list["ServiceReminder"]] = relationship( # noqa: F821
        "ServiceReminder", back_populates="vehicle", lazy="selectin"
    )
    options: Mapped[list["VehicleOption"]] = relationship(  # noqa: F821
        "VehicleOption", back_populates="vehicle", lazy="selectin", cascade="all, delete-orphan"
    )
    colors: Mapped[list["VehicleColor"]] = relationship(  # noqa: F821
        "VehicleColor", back_populates="vehicle", lazy="selectin", cascade="all, delete-orphan"
    )
    wakala_scores: Mapped["VehicleWakalaScore | None"] = relationship(  # noqa: F821
        "VehicleWakalaScore", back_populates="vehicle", uselist=False, lazy="selectin", cascade="all, delete-orphan"
    )

    @property
    def images(self) -> list[dict]:
        active_listing = next((l for l in self.listings if l.status == "active"), None)
        # Fallback to draft/any listing if no active listing
        if not active_listing and self.listings:
            active_listing = self.listings[0]
            
        if active_listing and active_listing.images_urls:
            return [{"file_path": url} for url in active_listing.images_urls]
        return []

    def __repr__(self) -> str:
        return f"<Vehicle {self.brand} {self.model} ({self.year}) — {self.price} MAD>"
